cmd/client/client.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	secret := u.User.String()
+	if !strings.Contains(secret, ":") {
+		b, err := base64.StdEncoding.DecodeString(secret)
+		if err != nil {
+			b, err = base64.RawStdEncoding.DecodeString(u.User.String())
+			if err != nil {
+				return serverConfig{}, fmt.Errorf("invalid password in key: %v", err)
+			}
+		}
+		secret = string(bytes.TrimSpace(b))
+	}
+	p := strings.Split(secret, ":")
+	if len(p) != 2 {
+		return serverConfig{}, fmt.Errorf("invalid password in key")
+	}
+
+	return serverConfig{
+		host:   u.Hostname(),
+		port:   port,
+		cipher: p[0],
+		secret: p[1],
+	}, nil
+}
+
+func resolveHostPort(addr string) (string, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return "", err
+	}
+	ip, err := net.ResolveIPAddr("ip", host)
+	if err != nil {
+		return "", fmt.Errorf("resolving hostname failed: %v", err)
+	}
+	return net.JoinHostPort(ip.String(), port), err
+}
+
+type SocksSSClient struct {
+	config   serverConfig
+	listener *net.TCPListener
+}
+
+// RunSocksSSClient starts a SOCKS server which proxies connections to the specified shadowsocks server.
+func RunSocksSSClient(bindAddr string, listenPort int, config serverConfig) (*SocksSSClient, error) {
+	listener, err := net.ListenTCP("tcp", &net.TCPAddr{IP: net.ParseIP(bindAddr), Port: listenPort})
+	if err != nil {
+		return nil, fmt.Errorf("listenTCP failed: %v", err)
+	}
+	logger.Infof("Listenting at %v", listener.Addr())
+
+	d, err := client.NewClient(config.host, config.port, config.secret, config.cipher)
+	if err != nil {
+		return nil, fmt.Errorf("failed connecting to server: %v", err)
+	}
+
+	go func() {
+		for {
+			clientConn, err := listener.AcceptTCP()
+			if err != nil {
+				if errors.Is(err, net.ErrClosed) {
+					logger.Info("SOCKS listener closed")
+				} else {
+					logger.Errorf("Accepting SOCKS connection failed: %v\n", err)
+				}
+				break
+			}
+			go func() {
+				defer clientConn.Close()
+
+				tgtAddr, err := socks.Handshake(clientConn)
+				if err != nil {
+					logger.Errorf("SOCKS handshake failed: %v", err)
+					return
+				}
+				addr, err := resolveHostPort(tgtAddr.String())
+				if err != nil {
+					logger.Errorf("Failed to resolve target address: %v", err)
+					return
+				}
+
+				logger.Debugf("Opening connection for %s", addr)
+				targetConn, err := d.DialTCP(nil, addr)
+				if err != nil {
+					logger.Errorf("Failed to dial: %v", err)
+					return
+				}
+				defer targetConn.Close()
+				_, _, err = onet.Relay(clientConn, targetConn)
+				if err != nil {
+					logger.Errorf("Relay failed: %v", err)
+					return
+				}
+				logger.Debugf("Connection closed %s", addr)
+			}()
+		}
+	}()
+	return &SocksSSClient{listener: listener}, nil
+}
+
+// ListenAddr returns the listening address used by the SOCKS server
+func (s *SocksSSClient) ListenAddr() net.Addr {
+	return s.listener.Addr()
+}
+
+// Stop stops the SOCKS server
+func (s *SocksSSClient) Stop() error {
+	return s.listener.Close()
+}
+
+func main() {
```

Comment:
Could you create a README.md file explaining how to run the socks proxy? Mention it only supports TCP.

Reply:
Would you rather have a separate README in `cmd/client` or do you think I should update the original README to change the `go-shadowsocks2` client to this one?